src/Init/Data/Array/Basic.lean

```diff
@@ -696,21 +705,26 @@ def erase [BEq α] (as : Array α) (a : α) : Array α :=
   | none   => as
   | some i => as.feraseIdx i
 
-def insertAtAux (i : Nat) (as : Array α) (j : Nat) : Array α :=
-  if h : i < j then
-    let as := as.swap! (j-1) j;
-    insertAtAux i as (j-1)
-  else
-    as
-
-/--
-  Insert element `a` at position `i`.
-  Pre: `i < as.size` -/
-def insertAt (as : Array α) (i : Nat) (a : α) : Array α :=
-  if i > as.size then panic! "invalid index"
-  else
-    let as := as.push a;
-    as.insertAtAux i as.size
+/-- Insert element `a` at position `i`. -/
+@[inline] def insertAt (as : Array α) (i : Fin (as.size + 1)) (a : α) : Array α :=
+  let rec loop (as : Array α) (j : Fin as.size) :=
+    if i.1 < j then
+      let j' := ⟨j-1, Nat.lt_of_le_of_lt (Nat.pred_le _) j.2⟩
+      let as := as.swap j' j
+      loop as ⟨j', by rw [size_swap]; exact j'.2⟩
+    else
+      as
+  let j := as.size
+  let as := as.push a
+  loop as ⟨j, size_push .. ▸ j.lt_succ_self⟩
+termination_by _ _ _ j => j.1
```

Comment:
You don't need to "fill" with `_`s. The given names are processed from right to left. It is not a bug but an intentional feature. Otherwise, we would have to calculate the free variables in the nested function.
That is, you can write just 
```lean
termination_by _ j => j.1
```
or 
```lean
termination_by loop j => j.1
```

Reply:
Do you have a style preference on using `termination_by loop` vs `termination_by _`? I almost always use `termination_by _` since it's rare for more than one definition in a mutual block or def + where to require a nontrivial termination proof, but arguably it's less clear what it's applying to.

Reply:
> Do you have a style preference on using termination_by loop vs termination_by _?

I (try to) use `termination_by loop` whenever more than one function is defined in the command.
However, I cannot claim I used this style consistently in the code base :)